File: src/main/model/DataChannels.java
package main.model;

public enum DataChannels {
	POSITION3("in_Position", 3, 0, 0),
	COLOR("in_Color", 3, 1, POSITION3.siB()),
	TEXCOORD("in_TextureCoord", 2, 2, POSITION3.siB() + COLOR.siB()),
	NORMAL("in_Normal", 3, 3, POSITION3.siB() + COLOR.siB() + TEXCOORD.siB()),
	BINORMAL("in_Binormal", 3, 4, POSITION3.siB() + COLOR.siB() + TEXCOORD.siB() + NORMAL.siB()),
	TANGENT("in_Tangent", 3, 5, POSITION3.siB() + COLOR.siB() + TEXCOORD.siB() + NORMAL.siB() +  BINORMAL.siB());

	private String binding;
	private int size;
	private int location;
	private int offset;

	DataChannels(String binding, int size, int location, int bufferOffset) {
		this.binding = binding;
		this.size = size;
		this.location = location;
		this.offset = bufferOffset;
	}
	
	public int getSize() {
		return size;
	}
	
	// Size in Bytes per Attribute
	public int siB() {
		return 4 * size;
	}

	public int getLocation() {
		return location;
	}

	public int getOffset() {
		return offset / 4;
	}
	public int getByteOffset() {
		return offset;
	}
	
	public String getBinding() {
		return binding;
	}
	
	@Override
	public String toString() {
		return String.format("%s (%d)", getBinding(), size); 
	}
}

File: engine/src/main/kotlin/de/hanno/hpengine/graphics/vertexbuffer/DataChannels.kt
package de.hanno.hpengine.graphics.vertexbuffer

import java.util.*

enum class DataChannels(//TODO Use not only float byte size
    val binding: String, val size: Int, val location: Int, val byteOffset: Int
) {
    POSITION3("in_Position", 3, 0, 0), COLOR("in_Color", 3, 1, POSITION3.sizeInBytes()), TEXCOORD(
        "in_TextureCoord",
        2,
        2,
        POSITION3.sizeInBytes() + COLOR.sizeInBytes()
    ),
    NORMAL("in_Normal", 3, 3, POSITION3.sizeInBytes() + COLOR.sizeInBytes() + TEXCOORD.sizeInBytes()), WEIGHTS(
        "in_Weights",
        4,
        5,
        POSITION3.sizeInBytes() + COLOR.sizeInBytes() + TEXCOORD.sizeInBytes() + NORMAL.sizeInBytes()
    ),
    JOINT_INDICES(
        "in_JointIndices",
        4,
        6,
        POSITION3.sizeInBytes() + COLOR.sizeInBytes() + TEXCOORD.sizeInBytes() + NORMAL.sizeInBytes() + WEIGHTS.sizeInBytes()
    );

    // Size in Bytes per Attribute
    fun sizeInBytes(): Int = java.lang.Float.BYTES * size

    fun getOffset(): Int = byteOffset / 4

    override fun toString(): String = String.format("%s (%d)", binding, size)

    companion object {
        fun totalElementsPerVertex(channels: EnumSet<DataChannels>): Int {
            var count = 0
            for (channel in channels) {
                count += channel.size
            }
            return count
        }

        fun bytesPerVertex(channels: EnumSet<DataChannels>): Int {
            var sum = 0
            for (channel in channels) {
                sum += channel.size
            }
            return sum * java.lang.Float.BYTES
        }
    }
}